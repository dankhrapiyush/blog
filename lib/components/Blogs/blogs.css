.clearfix:after {
    content: '';
    display: table;
    clear: both;
}

.root {
    padding: 2.5rem 0;
    margin: 0 auto;
}

.errorText {
    color: rgb(var(--venia-global-color-error));
    line-height: var(--venia-global-lineHeight-300);
}

.body {
    display: grid;
    gap: 3rem;
    grid-template-columns: 1fr 16rem;
    /* The summary grid item spans the entire right column. */
    grid-template-areas: 'items summary';
}

.itemsContainer {
    grid-area: items;
}

.summaryContainer {
    grid-area: summary;
    height: 100%;
}

.title {
    font-size: 1.5rem;
    font-weight: 400;
    padding: 0 0.5rem;
}

.pagination {
    position: relative;
    bottom: 0;
}

.blogTitle {
    color: rgb(var(--venia-text));
    padding-bottom: 1rem;
    font-size: 1.375rem;
    font-weight: 300;
    line-height: 1.375rem;
    text-align: center;
}

.postList {
    margin-top: 0px;
    margin-bottom: 20px;
    list-style: none;
}
.postList .postHeader .postTitle {
    max-width: 80%;
}
.postList .postHeader .postTitle a {
    text-decoration: none;
    font-size: 22px;
    color: #636363;
}
.postList .postHeader .postTitle {
    margin: 0;
    padding: 10px 0;
}
.postList .postContent .postReadMore {
    padding-top: 10px;
    float: right;
    color: #006bb4;
}

.postHolder {
    margin-bottom: 40px;
    padding: 20px;
    border: 1px solid #e9e9e9;
    border-radius: 5px;
}
.postHolder .postHeader .postInfo {
    padding: 0 0 20px 0;
}
.postHolder .postHeader .postInfo .item {
    display: inline-block;
}
.postHolder .postHeader .postInfo .dash {
    display: none;
}
.postHolder .postHeader .postInfo .label {
    font-weight: 700;
}
.postHolder .postHeader .postInfo .value {
    font-style: italic;
}
.postHolder .postHeader .postInfo .item::after {
    display: inline-block;
    content: ' | ';
    color: #b9b9b9;
    margin: 0 10px;
}
.postHolder .postHeader .postInfo .item:last-child::after {
    display: none;
}

/*
 * Mobile-specific styles.
 */

@media (max-width: 960px) {
    .root {
        padding-left: 1.5rem;
        padding-right: 1.5rem;
    }

    .body {
        /* Only one column in mobile view. */
        grid-template-columns: 100%;
        grid-template-areas:
            'items'
            'summary';
    }
}
