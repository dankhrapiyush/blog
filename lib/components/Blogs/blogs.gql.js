import gql from 'graphql-tag';

const GET_BLOG_QUERY = gql`
    query GetPosts($pageSize: Int, $currentPage: Int) {
        blogPosts(pageSize: $pageSize, currentPage: $currentPage) {
            total_count
            total_pages
            items {
                post_id
                identifier
                title
                content
                short_content
            }
        }
    }
`;

export default {
    queries: {
        getBlogQuery: GET_BLOG_QUERY
    },
    mutations: {}
};