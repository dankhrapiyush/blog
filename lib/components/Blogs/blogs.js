import React, { Fragment, useEffect } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { Title } from '@magento/venia-ui/lib/components/Head';
import { fullPageLoadingIndicator } from '@magento/venia-ui/lib/components/LoadingIndicator';
import { usePagination } from '@magento/peregrine';
import { useLazyQuery, useQuery } from '@apollo/client';

import BlogsContent from './blogsContent';
import defaultClasses from './blogs.css';

import GET_BLOG_POSTS from './queries/blogPosts.graphql';

const Blogs = props => {
    const { pageSize } = props;
    const { formatMessage } = useIntl();
    const classes = mergeClasses(defaultClasses, props.classes);

    const [paginationValues, paginationApi] = usePagination();
    const { currentPage, totalPages } = paginationValues;
    const { setCurrentPage, setTotalPages } = paginationApi;
    const pageControl = {
        currentPage,
        setPage: setCurrentPage,
        totalPages
    };

    // Scroll to the top on component load
    useEffect(() => {
        window.scrollTo(0, 0);
    }, []);

    // Fetch the data using apollo react hooks
    // https://www.apollographql.com/docs/react/api/react-apollo/
    // Search 'Valid fetchPolicy values are'
    // cache-first, cache-and-network, network-only, cache-only, no-cache
    // const { data, error, loading } = useQuery(getBlogQuery, { fetchPolicy: 'cache-and-network' });
    const [runQuery, queryResponse] = useLazyQuery(
        GET_BLOG_POSTS,
        {
            variables:
            {
                currentPage: Number(1),
                pageSize: Number(5)
            },
            fetchPolicy: 'cache-and-network'
        }
    );
    const { loading, error, data } = queryResponse;

    const totalPagesFromData = data
        ? data.blogPosts.total_pages
        : null;

    useEffect(() => {
        setTotalPages(totalPagesFromData);
        return () => {
            setTotalPages(null);
        };
    }, [setTotalPages, totalPagesFromData]);

    // Run the posts query immediately and whenever its variable values change.
    useEffect(() => {
        runQuery({
            variables: {
                currentPage: Number(currentPage),
                pageSize: Number(pageSize)
            },
            fetchPolicy: 'cache-and-network'
        });
        window.scrollTo({
            left: 0,
            top: 0,
            behavior: 'smooth'
        });
    }, [
        currentPage,
        pageSize,
        runQuery
    ]);

    // If we get an error after loading we should try to reset to page 1.
    // If we continue to have errors after that, render an error message.
    useEffect(() => {
        if (error && !loading && currentPage !== 1) {
            setCurrentPage(1);
        }
    }, [currentPage, error, loading, setCurrentPage]);

    // Display an error message.
    if (error && currentPage === 1 && !loading) return <div>Data Fetch Error</div>;

    // Show the loading indicator until data has been fetched.
    if (loading) return fullPageLoadingIndicator;
    const title = formatMessage(
        {
            id: 'blogs.title',
            defaultMessage: 'Blog'
        }
    );

    return (
        <Fragment>
            <Title>{title}</Title>
            <article className={classes.root}>
                <h1 className={classes.title}>
                    <div className={classes.blogTitle}>
                        <FormattedMessage
                            id={'blogs.title'}
                            defaultMessage={'Blog'}
                        />
                    </div>
                </h1>
            </article>
            <BlogsContent
                classes={classes}
                data={!loading && !data ? null : data.blogPosts}
                pageControl={pageControl}
            />
        </Fragment>
    );
};

export default Blogs;

Blogs.defaultProps = {
    pageSize: 5
};
