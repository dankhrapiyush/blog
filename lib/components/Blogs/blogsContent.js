import React, { Fragment } from 'react';
import { shape, arrayOf, number, string } from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import Pagination from '@magento/venia-ui/lib/components/Pagination';
import RichContent from '@magento/venia-ui/lib/components/RichContent';
import resourceUrl from '@magento/peregrine/lib/util/makeUrl';
import { Link } from 'react-router-dom';
import { intersperse } from '../../util/intersperse';
import { formatDate } from '../../util/formatDate';

import defaultClasses from './blogs.css';

const BlogsContent = props => {
    const { data, pageControl } = props;
    const classes = mergeClasses(defaultClasses, props.classes);

    const listItems = data && data.items ? data.items.map(blog => {
        const { post_id: id, identifier, title, content, short_content, publish_time, author, categories, tags } = blog;
        const linkTo = resourceUrl(`/blog/post/${String(identifier)}`);

        const categoryList = categories.map((category) => {
                const categoryLink = resourceUrl(`/blog/category/${category.identifier}`);
                return category.is_active === 1 &&
                    <Link to={categoryLink} key={category.category_id}>{category.title}</Link>
            }
        );

        const tagList = tags.map((tag) => {
                const tagLink = resourceUrl(`/blog/tag/${tag.identifier}`);
                return tag.is_active === 1 &&
                    <Link to={tagLink} key={tag.tag_id}>{tag.title}</Link>
            }
        );

        const blogDescription = short_content ? (
            <RichContent html={short_content} />
        ) : <RichContent html={content} />;

        return (
            <li key={id} className={classes.postHolder}>
                <div className={classes.postHeader}>
                    <div className={classes.postTitleHolder}>
                        <h2 className={classes.postTitle}>
                            <Link to={linkTo}>{title}</Link>
                        </h2>
                    </div>
                    <div className={classes.postInfo}>
                        <div className={classes.item}>
                            <span className={classes.label}>
                                <FormattedMessage
                                    id={'blogsContent.posted'}
                                    defaultMessage={'Posted: '}
                                />
                            </span>
                            <span className={classes.value}>
                                {formatDate(publish_time)}
                            </span>
                        </div>
                        {categoryList.length > 0 &&
                            <div className={classes.item}>
                                <span className={classes.label}>
                                    <FormattedMessage
                                        id={'blogsContent.categories'}
                                        defaultMessage={'Categories: '}
                                    />
                                </span>
                                <span className={classes.value}>
                                    {intersperse(categoryList, ', ')}
                                </span>
                            </div>
                        }
                        {tagList.length > 0 &&
                            <div className={classes.item}>
                                <span className={classes.label}>
                                    <FormattedMessage
                                        id={'blogsContent.tags'}
                                        defaultMessage={'Tags: '}
                                    />
                                </span>
                                <span className={classes.value}>
                                    {intersperse(tagList, ', ')}
                                </span>
                            </div>
                        }
                        <div className={classes.item}>
                            <span className={classes.label}>
                                <FormattedMessage
                                    id={'blogsContent.author'}
                                    defaultMessage={'Author: '}
                                />
                            </span>
                            <span className={classes.value}>{author.name}</span>
                        </div>
                    </div>
                </div>
                <div className={classes.postContent}>
                    <div className={classes.clearfix}>
                        <div className={classes.clearfix}>
                            {blogDescription}
                        </div>
                        <Link className={classes.postReadMore} to={linkTo}>
                            <FormattedMessage
                                id={'blogsContent.readMore'}
                                defaultMessage={'Read more'}
                            /> &#187;
                        </Link>
                    </div>
                </div>
            </li>
        );
    }) : '';

    const content =
        data && data.total_count === 0 ? (
                <p>
                    <FormattedMessage
                        id={'blogsContent.notFound'}
                        defaultMessage={'Sorry! We couldn\'t find any blogs.'}
                    />
                </p>
            ) : (
                <Fragment>
                    <ul className={classes.postList}>{listItems}</ul>
                    <div className={classes.pagination}>
                        <Pagination pageControl={pageControl} />
                    </div>
                </Fragment>
            );

    return <Fragment>{content}</Fragment>;
};

BlogsContent.prototype = {
    data: shape({
        items: arrayOf(
            shape({
                id: number.isRequired,
                identifier: string.isRequired,
                title: string.isRequired,
                content: string,
                short_content: string,
                publish_time: string,
                author: shape({
                    name: string.isRequired
                }),
                categories: arrayOf(
                    shape({
                        category_id: number.isRequired,
                        identifier: string.isRequired,
                        title: string.isRequired,
                        is_active: number.isRequired
                    })
                ),
                tags: arrayOf(
                    shape({
                        tag_id: number.isRequired,
                        identifier: string.isRequired,
                        title: string.isRequired,
                        is_active: number.isRequired
                    })
                )
            })
        )
    }).isRequired
};

export default BlogsContent;
