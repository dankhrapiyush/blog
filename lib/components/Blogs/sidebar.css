.block {
    margin-bottom: 30px;
}
.blockTitle {
    font-size: 1rem;
    color: rgb(var(--venia-global-color-gray-900));
    font-weight: 600;
    line-height: 1.5;
    margin: 0 0 10px;
}
.list {
    list-style: disc;
}
.list li {
    margin: 5px 0 5px 28px;
}
