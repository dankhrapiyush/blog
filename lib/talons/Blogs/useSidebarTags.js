import { useQuery } from '@apollo/client';

/**
 * Talon for Blog Sidebar Tags.
 *
 * @param {*} props.query the blog data query
 */
export const useSidebarTags = props => {
    const { query } = props;

    const { error, loading, data } = useQuery(query, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first'
    });

    let items = [];
    if (!error && !loading) {
        items = data.blogTags.items;
    }

    const hasError = !!error;
    const total_count = data ? data.blogTags.total_count : 0;

    return {
        hasError,
        items,
        total_count
    };
};
