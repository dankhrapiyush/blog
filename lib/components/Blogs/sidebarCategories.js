import React, { Fragment } from 'react';
import { shape, arrayOf, number, string } from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { useStyle } from '@magento/venia-ui/lib/classify';
import { Link } from 'react-router-dom';
import resourceUrl from '@magento/peregrine/lib/util/makeUrl';
import { useSidebarCategories } from '../../talons/Blogs/useSidebarCategories';

import defaultClasses from './sidebar.css';
import { gql } from '@apollo/client';

const SidebarCategories = props => {
    const talonProps = useSidebarCategories({
        query: GET_BLOG_CATEGORIES
    });
    const { total_count, items } = talonProps;

    if (!total_count || !items.length) {
        return null;
    }

    const classes = useStyle(defaultClasses, props.classes);

    const categoryList = items.map((category) => {
        const categoryLink = resourceUrl(
            `/blog/category/${category.identifier}`
        );
        return (
            category.is_active === 1 && (
                <li key={category.category_id}>
                    <Link to={categoryLink}>
                        {category.title} {`(${category.posts_count})`}
                    </Link>
                </li>
            )
        );
    });

    return (
        <Fragment>
            <div className={classes.block}>
                <div className={classes.blockTitle}>
                    <FormattedMessage
                        id={'sidebarCategories.title'}
                        defaultMessage={'Categories'}
                    />
                </div>
                <ul className={classes.list}>{categoryList}</ul>
            </div>
        </Fragment>
    );
};

SidebarCategories.prototype = {
    data: shape({
        items: arrayOf(
            shape({
                category_id: number.isRequired,
                identifier: string.isRequired,
                title: string.isRequired,
                is_active: number,
                posts_count: number
            })
        )
    }).isRequired
};

export const GET_BLOG_CATEGORIES = gql`
    query blogCategories {
        blogCategories {
            total_count
            items {
                category_id
                identifier
                title
                is_active
                posts_count
            }
        }
    }
`;

export default SidebarCategories;
