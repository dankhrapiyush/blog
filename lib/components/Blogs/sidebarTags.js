import React, { Fragment } from 'react';
import { shape, arrayOf, number, string } from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { Link, resourceUrl } from '@magento/venia-drivers';
import { useSidebarTags } from '../../talons/Blogs/useSidebarTags';

import defaultClasses from './sidebar.css';
import { gql } from '@apollo/client';

const SidebarTags = props => {
    const talonProps = useSidebarTags({
        query: GET_BLOG_TAGS
    });
    const { total_count, items } = talonProps;

    if (!total_count || !items.length) {
        return null;
    }

    const classes = mergeClasses(defaultClasses, props.classes);

    const tagList = items.map((tag) => {
        const tagLink = resourceUrl(`/blog/tag/${tag.identifier}`);
        return (
            tag.is_active === 1 && (
                <li key={tag.tag_id}>
                    <Link to={tagLink}>{tag.title}</Link>
                </li>
            )
        );
    });

    return (
        <Fragment>
            <div className={classes.block}>
                <div className={classes.blockTitle}>
                    <FormattedMessage
                        id={'sidebarTags.title'}
                        defaultMessage={'Tags'}
                    />
                </div>
                <ul className={classes.list}>{tagList}</ul>
            </div>
        </Fragment>
    );
};

SidebarTags.prototype = {
    data: shape({
        items: arrayOf(
            shape({
                tag_id: number.isRequired,
                identifier: string.isRequired,
                title: string.isRequired,
                is_active: number
            })
        )
    }).isRequired
};

export const GET_BLOG_TAGS = gql`
    query blogTags {
        blogTags {
            total_count
            items {
                tag_id
                identifier
                title
                is_active
            }
        }
    }
`;

export default SidebarTags;
