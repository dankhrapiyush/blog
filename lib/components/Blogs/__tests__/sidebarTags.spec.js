import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { default as createTestInstance } from '@magento/peregrine/lib/util/createTestInstance';
import { useSidebarTags } from '../../../talons/Blogs/useSidebarTags';
import SidebarTags from '../sidebarTags';
import { IntlProvider } from 'react-intl';

jest.mock('@magento/venia-ui/lib/classify');
jest.mock('@magento/venia-drivers', () => ({
    Link: ({ children, ...rest }) => <div {...rest}>{children}</div>,
    resourceUrl: x => x
}));
jest.mock('../../../talons/Blogs/useSidebarTags', () => ({
    useSidebarTags: jest.fn()
}));

const talonProps = {
    total_count: 2,
    items: [
        {
            tag_id: 1,
            identifier: 'apple',
            title: 'apple',
            is_active: 1
        },
        {
            tag_id: 2,
            identifier: 'ios',
            title: 'ios',
            is_active: 1
        },
        {
            tag_id: 3,
            identifier: 'ipad',
            title: 'ipad',
            is_active: 1
        },
        {
            tag_id: 4,
            identifier: 'news',
            title: 'news',
            is_active: 1
        }
    ]
};

test('renders sidebar tags list', () => {
    useSidebarTags.mockReturnValueOnce(talonProps);

    const root = createTestInstance(
        <MemoryRouter>
            <IntlProvider locale="en-US">
                <SidebarTags />
            </IntlProvider>
        </MemoryRouter>
    );

    expect(root.toJSON()).toMatchSnapshot();
});

test('renders sidebar tags list when count is zero', () => {
    useSidebarTags.mockReturnValueOnce({
        total_count: 0
    });

    const root = createTestInstance(
        <MemoryRouter>
            <IntlProvider locale="en-US">
                <SidebarTags />
            </IntlProvider>
        </MemoryRouter>
    );

    expect(root.toJSON()).toMatchSnapshot();
});

test('renders sidebar tags list when count is zero and data exist', () => {
    useSidebarTags.mockReturnValueOnce({
        ...talonProps,
        total_count: 0
    });

    const root = createTestInstance(
        <MemoryRouter>
            <IntlProvider locale="en-US">
                <SidebarTags />
            </IntlProvider>
        </MemoryRouter>
    );

    expect(root.toJSON()).toMatchSnapshot();
});

test('renders sidebar tags list when count is greater than zero and no data exist', () => {
    useSidebarTags.mockReturnValueOnce({
        ...talonProps,
        items: []
    });

    const root = createTestInstance(
        <MemoryRouter>
            <IntlProvider locale="en-US">
                <SidebarTags />
            </IntlProvider>
        </MemoryRouter>
    );

    expect(root.toJSON()).toMatchSnapshot();
});
