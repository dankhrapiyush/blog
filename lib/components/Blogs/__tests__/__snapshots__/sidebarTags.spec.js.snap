// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders sidebar tags list 1`] = `
<div
  className="block"
>
  <div
    className="blockTitle"
  >
    <mock-FormattedMessage
      defaultMessage="Tags"
      id="sidebarTags.title"
    />
  </div>
  <ul
    className="list"
  >
    <li>
      <div
        to="/blog/tag/apple"
      >
        apple
      </div>
    </li>
    <li>
      <div
        to="/blog/tag/ios"
      >
        ios
      </div>
    </li>
    <li>
      <div
        to="/blog/tag/ipad"
      >
        ipad
      </div>
    </li>
    <li>
      <div
        to="/blog/tag/news"
      >
        news
      </div>
    </li>
  </ul>
</div>
`;

exports[`renders sidebar tags list when count is greater than zero and no data exist 1`] = `null`;

exports[`renders sidebar tags list when count is zero 1`] = `null`;

exports[`renders sidebar tags list when count is zero and data exist 1`] = `null`;
